src/riak_ensemble_peer.erl

```diff
@@ -1850,12 +1863,17 @@ handle_event({watch_leader_status, Pid}, StateName, State) when node(Pid) =/= no
     {next_state, StateName, State};
 handle_event({watch_leader_status, Pid}, StateName, State = #state{watchers = Watchers}) ->
     _ = notify_leader_status(Pid, StateName, State),
-    %% Might as well take this opportunity to prune any dead pids that are in the list
-    NewWatcherList = [P || P <- [Pid | Watchers], is_process_alive(P)],
-    {next_state, StateName, State#state{watchers = NewWatcherList}};
+    MRef = erlang:monitor(process, Pid),
+    {next_state, StateName, State#state{watchers = [{Pid, MRef} | Watchers]}};
 handle_event({stop_watching, Pid}, StateName, State = #state{watchers = Watchers}) ->
-    NewWatcherList = lists:delete(Pid, Watchers),
-    {next_state, StateName, State#state{watchers = NewWatcherList}};
+    case lists:keytake(Pid, 1, Watchers) of
+        false ->
+            lager:warning("Tried to stop watching for pid ~p, but did not find it in watcher list"),
```

Comment:
Not sure if this shouldn't be a debug-level thing, rather than warning. In an operational sense, is anyone going to care if this happens in their cluster?